[OpenACC] Implement Atomic construct variants

erichkeane opened the pull request:
`atomic` is required to be followed by a special `atomic clause`, so this patch manages the parsing of that.  We are representing each of the variants of the atomic construct as separate kinds, because they have distinct rules/application/etc, and this should make it easier to check rules in the future.

erichkeane committed:
[OpenACC] Implement Atomic construct variants

`atomic` is required to be followed by a special `atomic clause`, so
this patch manages the parsing of that.  We are representing each of the
variants of the atomic construct as separate kinds, because they have
distinct rules/application/etc, and this should make it easier to check
rules in the future.

alexey-bataev commented on clang/lib/Parse/ParseOpenACC.cpp:
```diff
@@ -59,9 +60,21 @@ OpenACCDirectiveKindEx getOpenACCDirectiveKind(StringRef Name) {
+// Since 'atomic' is effectively a compound directive, this will decode the
+// second part of the directive.
+OpenACCDirectiveKind getOpenACCAtomicDirectiveKind(StringRef Name) {
+  return llvm::StringSwitch<OpenACCDirectiveKind>(Name)
+      .Case("read", OpenACCDirectiveKind::AtomicRead)
+      .Case("write", OpenACCDirectiveKind::AtomicWrite)
+      .Case("update", OpenACCDirectiveKind::AtomicUpdate)
+      .Case("capture", OpenACCDirectiveKind::AtomicCapture)
```
Are this directive kinds or clauses?

erichkeane replied:
So they are considered in the language to be 'atomic clauses', but they act more like constructs/directives, in that they have their own rules as far as appertainment and other clause legality.

Note they aren't listed as a 'clause' in the standard like the REST of the 'clauses', so they are kind of a special case that applies to atomic only.

Since atomic REQUIRES this 'atomic-clause' (again, clause is misleading here IMO), and they are so different, it seems to make the most sense to me to represent them as their own directives here.

So to answer your question succinctly: "They are basically neither".

alexey-bataev replied:
I think this may cause problems in future, if new combinations of clauses will appear. It will be very hard to redesign it to represent the new clauses as directives-kind (especially if the new clauses will support non-strict use order).

erichkeane replied:
The list is a pretty complete set, right? These haven't changed in a very long time as far as I can tell, so I considered the risk of that to be minimal.

alexey-bataev replied:
It is very similar to OpenMP, which was extended recently. I assume OpenACC also may want to add some new features here later

alexey-bataev commented on clang/test/ParserOpenACC/parse-constructs.c:
```diff
@@ -94,6 +94,37 @@ void func() {
   for(;;){}
 
+  int i = 0, j = 0, k = 0;
+  // expected-error@+2{{missing OpenACC 'atomic-clause'; expected 'read', 'write', 'update', or 'capture'}}
+  // expected-warning@+1{{OpenACC directives not yet implemented, pragma ignored}}
+#pragma acc atomic
```
It is allowed by the standard, means same as atomic update

erichkeane replied:
Ah!  I'd not caught that on my read through the standard.  Thanks!

erichkeane committed:
Refactor to no longer consider atomic variants as constructs

After review, it was pointed out that this is a list that has changed
recently in OpenMP and thus might change here.

This patch extracts the parsing of atomic-clause to happen after
directive determination.

erichkeane commented on clang/lib/Parse/ParseOpenACC.cpp:
```diff
@@ -143,26 +139,25 @@ ParseOpenACCEnterExitDataDirective(Parser &P, Token FirstTok,
+  // this as a clause list, which, despite being invalid, is likely what the
+  // user was trying to do.
+  if (AtomicKind == OpenACCAtomicKind::Invalid)
+    return OpenACCAtomicKind::Update;
```
I don't believe so, since this is not diagnosing on it, I'm treating it as just `#pragma acc atomic` and treating the 'rest' as a 'clause-list'.  That way we can diagnose the rest as a 'clauses not allowed here'.
